File: examples/ffi/rust_calling_c/c/matrix.c
// Copyright 2017 The Bazel Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ffi/rust_calling_c/c/matrix.h"

#include <stdio.h>
#include <string.h>

Matrix* matrix_new(size_t rows, size_t cols, const uint64_t* data) {
  if (data == NULL) {
    return NULL;
  }
  Matrix* matrix = (Matrix*)malloc(sizeof(*matrix));
  if (matrix == NULL) {
    return NULL;
  }
  matrix->rows = rows;
  matrix->cols = cols;
  matrix->data = (uint64_t*)malloc(rows * cols * sizeof(*(matrix->data)));
  memcpy(matrix->data, data, rows * cols * sizeof(*data));
  return matrix;
}

int matrix_at(const Matrix* matrix, size_t row, size_t col, uint64_t* n) {
  if (matrix == NULL || matrix->data == NULL || n == NULL) {
    return 0;
  }
  if (row >= matrix->rows || col >= matrix->cols) {
    return 0;
  }
  *n = matrix->data[row * matrix->cols + col];
  return 1;
}

int matrix_set(const Matrix* matrix, size_t row, size_t col, uint64_t n) {
  if (matrix == NULL || matrix->data == NULL) {
    return 0;
  }
  if (row >= matrix->rows || col >= matrix->cols) {
    return 0;
  }
  matrix->data[row * matrix->cols + col] = n;
  return 1;
}

void matrix_transpose(Matrix* matrix) {
  if (matrix == NULL || matrix->data == NULL) {
    return;
  }

  size_t len = matrix->rows * matrix->cols;
  int* visited = (int*)malloc(len * sizeof(*visited));
  if (visited == NULL) {
    return;
  }
  memset(visited, 0, len * sizeof(*visited));

  // Follow-the-cycles implementation of matrix transposition. Note that we
  // skip the last element since it always has a cycle of length 1 and thus
  // does not need to be moved.
  size_t q = matrix->rows * matrix->cols - 1;
  for (size_t i = 0; i < q; ++i) {
    if (visited[i] == 1) {
      continue;
    }
    size_t current_idx = i;
    size_t next_idx = i;
    do {
      visited[current_idx] = 1;
      next_idx = (current_idx * matrix->cols) % q;
      if (next_idx == i) {
        break;
      }

      uint64_t current_val = matrix->data[current_idx];
      matrix->data[current_idx] = matrix->data[next_idx];
      matrix->data[next_idx] = current_val;
      current_idx = next_idx;
    } while (1);
  }

  free(visited);
  size_t cols = matrix->rows;
  matrix->rows = matrix->cols;
  matrix->cols = cols;
}

int matrix_equal(const Matrix* a, const Matrix* b) {
  if (a == NULL || b == NULL || a->data == NULL || b->data == NULL) {
    return 0;
  }
  if (a->rows != b->rows || a->cols != b->cols) {
    return 0;
  }
  size_t len = a->rows * a->cols;
  for (size_t i = 0; i < len; ++i) {
    if (a->data[i] != b->data[i]) {
      return 0;
    }
  }
  return 1;
}

void matrix_free(Matrix* matrix) {
  if (matrix == NULL) {
    return;
  }
  if (matrix->data != NULL) {
    free(matrix->data);
  }
  free(matrix);
}

File: examples/ffi/rust_calling_c/src/matrix.rs
// Copyright 2017 The Bazel Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

extern crate libc;

mod ffi;

use std::ops;
use std::ptr;

/// Wrapper around pointer to FFI Matrix struct.
pub struct Matrix {
    matrix: *mut ffi::Matrix,
}

/// Wrapper around low-level FFI Matrix API.
impl Matrix {
    /// Constructs a new Matrix from the given data. Matrix returned contains a copy of the data
    /// provided.
    ///
    /// # Panics
    ///
    /// If rows * cols does not equal data.len() or if matrix could not be allocated.
    pub fn new(rows: usize, cols: usize, data: &[u64]) -> Matrix {
        if data.len() != rows * cols {
            panic!(
                "rows * cols ({}) do not equal data.len() ({})",
                rows * cols,
                data.len()
            );
        }

        let mut data_copy: Vec<u64> = vec![0; data.len()];
        data_copy.clone_from_slice(data);
        unsafe {
            let matrix: *mut ffi::Matrix = ffi::matrix_new(rows, cols, data_copy.as_ptr());
            if matrix.is_null() {
                panic!("Failed to allocate Matrix.");
            }
            Matrix { matrix }
        }
    }

    /// Fetches the value at the specified row and column.
    pub fn at(&self, row: usize, col: usize) -> Option<u64> {
        let mut n: u64 = 0;
        unsafe {
            if ffi::matrix_at(self.matrix, row, col, &mut n) == 0 {
                return None;
            }
        }
        Some(n)
    }

    /// Sets the value at the specified row and column.
    ///
    /// # Panics
    ///
    /// If row, col is out of bounds.
    pub fn set(&mut self, row: usize, col: usize, n: u64) {
        unsafe {
            if ffi::matrix_set(self.matrix, row, col, n) == 0 {
                panic!("Row {}, col {} is out of bounds.", row, col);
            }
        }
    }

    /// Returns the number of rows of the matrix.
    pub fn rows(&self) -> usize {
        unsafe { (*self.matrix).rows }
    }

    /// Returns the number of cols of the matrix.
    pub fn cols(&self) -> usize {
        unsafe { (*self.matrix).cols }
    }

    /// Performs an in-place transposition of the matrix.
    pub fn transpose(&mut self) {
        unsafe {
            ffi::matrix_transpose(self.matrix);
        }
    }

    /// Checks whether the matrix is equal to the provided Matrix.
    pub fn equal(&self, other: &Matrix) -> bool {
        unsafe { ffi::matrix_equal(self.matrix, other.matrix) != 0 }
    }
}

impl ops::Drop for Matrix {
    fn drop(&mut self) {
        unsafe {
            ffi::matrix_free(self.matrix);
        }
        self.matrix = ptr::null_mut();
    }
}

#[cfg(test)]
mod test {
    use super::Matrix;

    #[test]
    fn test_size() {
        let matrix = Matrix::new(2, 4, &[11, 12, 13, 14, 21, 22, 23, 24]);
        assert_eq!(2, matrix.rows());
        assert_eq!(4, matrix.cols());
    }

    #[test]
    fn test_equal() {
        let matrix_a = Matrix::new(2, 4, &[11, 12, 13, 14, 21, 22, 23, 24]);
        let matrix_b = Matrix::new(2, 4, &[11, 12, 13, 14, 21, 22, 23, 24]);
        assert!(matrix_a.equal(&matrix_b));

        let matrix_c = Matrix::new(2, 4, &[12, 13, 14, 15, 23, 24, 25, 26]);
        assert!(!matrix_a.equal(&matrix_c));
    }

    #[test]
    fn test_transpose() {
        let mut matrix = Matrix::new(2, 4, &[11, 12, 13, 14, 21, 22, 23, 24]);
        matrix.transpose();
        let expected = Matrix::new(4, 2, &[11, 21, 12, 22, 13, 23, 14, 24]);
        assert!(matrix.equal(&expected));
    }
}
